src/layers/transform/gather.cu

```diff
@@ -33,87 +33,107 @@ namespace lbann {
 namespace {
 
 using Dim2 = gpu_lib::array<size_t, 2>;
+using Dim3 = gpu_lib::array<size_t, 3>;
 
 /** @brief Kernel for scattering a 2D tensor along dim 1
  *
  *  output(j,indices(j,i)) = values(j,i)
  *
- *  Block dimensions: bdimx x bdimy x 1
+ *  Block dimensions: bdimx x bdimy x bdimz
  *
- *  Grid dimensions: (values_dim[1] / bdimx) x (values_dim[0] / bdimy) x 1
+ *  Grid dimensions: (num_columns_input_mat / bdimx) x (num_rows / bdimy) x mb_size /bdimz
  */
 template <typename T>
 __global__ void scatter2d_kernel(
   const T* __restrict__ indices,
   Dim2 indices_strides,
   const T* __restrict__ values,
-  Dim2 values_dims,
-  Dim2 values_strides,
+  Dim3 values_dims,
+  Dim3 values_strides,
   T* __restrict__ output,
-  Dim2 output_dims,
-  Dim2 output_strides) {
+  Dim3 output_dims,
+  Dim3 output_strides) {
 
   // Indices
   const size_t gidx = threadIdx.x + blockIdx.x * blockDim.x;
   const size_t gidy = threadIdx.y + blockIdx.y * blockDim.y;
+  const size_t gidz = threadIdx.z + blockIdx.z * blockDim.z;
+
   const size_t nthreadsx = gridDim.x * blockDim.x;
   const size_t nthreadsy = gridDim.y * blockDim.y;
-
-  for (size_t j=gidy; j<values_dims[0]; j+=nthreadsy) {
-    for (size_t i=gidx; i<values_dims[1]; i+=nthreadsx) {
-      const auto ind = static_cast<El::Int>(
-        gpu_lib::floor(
-          indices[j*indices_strides[0] + i*indices_strides[1]]));
-      if (0<=ind && ind<static_cast<El::Int>(output_dims[1])) {
-        const auto& x = values[j*values_strides[0] + i*values_strides[1]];
-        auto& y = output[j*output_strides[0] + ind*output_strides[1]];
-        gpu_lib::atomic_add(&y, x);
+  const size_t nthreadsz = gridDim.z * blockDim.z;
+
+  auto mini_batch_size = output_dims[0];
+  auto num_rows = output_dims[1];
+  auto num_out_columns = output_dims[2];
+  auto num_value_columns = values_dims[2];
+
+  for (size_t batch = gidz; batch < mini_batch_size; batch+=nthreadsz){
+    for(size_t row = gidy; row < num_rows; row+=nthreadsy){
+      for (size_t i = gidx; i < num_value_columns; i+=nthreadsx){
+        const auto ind = static_cast<El::Int>(
+          gpu_lib::floor(
+            indices[batch*indices_strides[0] + i*indices_strides[1]]));
+        if (0<=ind && ind < static_cast<El::Int>(num_out_columns)){
+          const auto& x = values[batch*values_strides[0] + row*values_strides[1] + i*values_strides[2]];
+          auto &y = output[batch*output_strides[0] + row*output_strides[1] + ind*output_strides[2]];
+          gpu_lib::atomic_add(&y, x);
+        }
       }
     }
   }
-
 }
 
 /** @brief Kernel for gathering a 2D tensor along dim 1
  *
  *  output(j,i) = values(j,indices(j,i))
  *
- *  Block dimensions: bdimx x bdimy x 1
+ *  Block dimensions: bdimx x bdimy x bdimz
  *
- *  Grid dimensions: (output_dim[1] / bdimx) x (output_dim[0] / bdimy) x 1
+ *  Grid dimensions: (num_columns_output_mat / bdimx) x (num_rows / bdimy) x mb_size /bdimz
```

Comment:
These grid dimensions aren't used when launching the kernel.

Reply:
Fixed!